File: Objective-c/KeyboardHandle.m
//
//  KeyboardHandle.m
//  keyboardHandle - Objc
//
//  Created by EVERTRUST on 2017/10/31.
//  Copyright © 2017年 EVERTRUST. All rights reserved.
//

#import "KeyboardHandle.h"
#import "UIResponder+UIResponder_Categroy.h"
#import <UIKit/UIKit.h>

#define SPACE 20.0
@implementation KeyboardHandle
{
    BOOL _enable;
}

-(instancetype)init {
    _enable = true;
    return self;
}

-(void)setIsEnable:(BOOL)isEnable {
    _enable = isEnable;
    if (isEnable) {
        [[NSNotificationCenter defaultCenter] removeObserver:self];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShow:) name:UIKeyboardWillShowNotification object:nil];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillhide:) name:UIKeyboardWillHideNotification object:nil];
    } else {
        [[NSNotificationCenter defaultCenter] removeObserver:self];
    }
}

-(void) keyboardWillShow: (NSNotification*) notification {
    NSValue* keyboardFrame = [notification userInfo][UIKeyboardFrameEndUserInfoKey];
    UIViewController* topVC = [self topVC:UIApplication.sharedApplication.keyWindow.rootViewController];
    int animationOption = [(NSNumber*)[notification userInfo][UIKeyboardAnimationCurveUserInfoKey] intValue];
    double duration = [[notification userInfo][UIKeyboardAnimationDurationUserInfoKey] doubleValue];
    NSLog(@"%f", duration);
    UIViewAnimationOptions options = animationOption<<16;
    UIView* first = (UIView*)[UIResponder firstResponder];
    if (first) {
        float rect = [first convertPoint:CGPointMake(0, first.frame.size.height) toView:topVC.view].y;
        float offsetX = rect + keyboardFrame.CGRectValue.size.height - topVC.view.frame.size.height + SPACE;
        if (offsetX > 0) {
            [UIView animateWithDuration:duration delay:0 options:options animations:^{
                topVC.view.frame = CGRectMake(0, -offsetX, topVC.view.frame.size.width, topVC.view.frame.size.height);
            } completion:nil];
        } else {
            [UIView animateWithDuration:duration delay:0 options:options animations:^{
                topVC.view.frame = CGRectMake(0, 0, topVC.view.frame.size.width, topVC.view.frame.size.height);
            } completion:nil];
            
        }
    }
}

-(void) keyboardWillhide: (NSNotification*) notification {
    UIViewController* topVC = [self topVC:UIApplication.sharedApplication.keyWindow.rootViewController];
    int animationOption = [(NSNumber*)[notification userInfo][UIKeyboardAnimationCurveUserInfoKey] intValue];
    UIViewAnimationOptions options = animationOption<<16;
    [UIView animateWithDuration:0.25 delay:0 options:options animations:^{
        topVC.view.frame = CGRectMake(0, 0, topVC.view.frame.size.width, topVC.view.frame.size.height);
    } completion:nil];

}


-(UIViewController*) topVC: (nullable UIViewController*) viewController {
    viewController = UIApplication.sharedApplication.keyWindow.rootViewController;

    if ([viewController isKindOfClass:[UINavigationController class]]) {
        UINavigationController* navCtrl = (UINavigationController*) viewController;
        return [self topVC:navCtrl.visibleViewController];
    }
    if ([viewController isKindOfClass:[UITabBarController class]]) {
        UITabBarController* tabCtrl = (UITabBarController*) viewController;
        return [self topVC:tabCtrl.selectedViewController];
    }
    if ([viewController presentedViewController]) {
        return [self topVC:[viewController presentedViewController]];
    }
    return viewController;
}

@end

File: Swift/KeyboardHandle.swift
//
//  KeyboardHandle.swift
//  HelloMyKeyboard
//
//  Created by EVERTRUST on 2017/10/31.
//  Copyright © 2017年 EVERTRUST. All rights reserved.
//

import UIKit

let keyboardHandle = KeyboardHandle()

final class KeyboardHandle {
    fileprivate var _enable = true
    var isEnable: Bool {
        get {
            return _enable
        } set {
            _enable = newValue
            if newValue {
                NotificationCenter.default.removeObserver(self)
                NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow(_:)), name: NSNotification.Name.UIKeyboardWillShow, object: nil)
                NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillhide(_:)), name: NSNotification.Name.UIKeyboardWillHide, object: nil)
            } else {
                NotificationCenter.default.removeObserver(self)
            }
        }
    }
    
    fileprivate init() {}
    
    @objc func keyboardWillShow(_ notification: Notification) {
        guard let keyboardFrame: NSValue = notification.userInfo?[UIKeyboardFrameEndUserInfoKey] as? NSValue else { return }
        guard let topVC = UIApplication.topViewController() else { return }
        guard let animationOption = (notification.userInfo?[UIKeyboardAnimationCurveUserInfoKey] as? NSNumber)?.intValue else { return }
        guard let duration: Double = (notification.userInfo?[UIKeyboardAnimationDurationUserInfoKey] as? NSNumber)?.doubleValue else { return }
        let options = UIViewAnimationOptions(rawValue: UInt(animationOption<<16))
        if let first = UIResponder.firstResponder() as? UIView {
            let rect = first.convert(CGPoint(x: 0, y: first.frame.height), to: topVC.view).y
            let offsetX = rect + keyboardFrame.cgRectValue.height - topVC.view.frame.height + 20
            if offsetX > 0 {
                UIView.animate(withDuration: duration, delay: 0, options: options, animations: {
                    topVC.view.frame = CGRect(x: 0.0, y: -offsetX, width: topVC.view.frame.width, height: topVC.view.frame.height)
                }, completion: nil)
            } else {
                UIView.animate(withDuration: duration, delay: 0, options: options, animations: {
                    topVC.view.frame = CGRect(x: 0, y: 0, width: topVC.view.frame.width, height: topVC.view.frame.height)
                }, completion: nil)
            }
            
        }
    }
    
    @objc func keyboardWillhide(_ notification: Notification) {
        guard let topVC = UIApplication.topViewController() else { return }
        guard let animationOption = (notification.userInfo?[UIKeyboardAnimationCurveUserInfoKey] as? NSNumber)?.intValue else { return }
        guard let duration: Double = (notification.userInfo?[UIKeyboardAnimationDurationUserInfoKey] as? NSNumber)?.doubleValue else { return }
        let options = UIViewAnimationOptions(rawValue: UInt(animationOption<<16))
        
        UIView.animate(withDuration: duration, delay: 0, options: options, animations: {
            topVC.view.frame = CGRect(x: 0, y: 0, width: topVC.view.frame.width, height: topVC.view.frame.height)
        }, completion: nil)
    }
}

extension UIApplication {
    class func topViewController(controller: UIViewController? = UIApplication.shared.keyWindow?.rootViewController) -> UIViewController? {
        if let navigationController = controller as? UINavigationController {
            return topViewController(controller: navigationController.visibleViewController)
        }
        if let tabController = controller as? UITabBarController {
            if let selected = tabController.selectedViewController {
                return topViewController(controller: selected)
            }
        }
        if let presented = controller?.presentedViewController {
            return topViewController(controller: presented)
        }
        return controller
    }
}

extension UIResponder {
    private weak static var _currentFirstResponder: UIResponder? = nil
    public class func firstResponder() -> UIResponder? {
        UIResponder._currentFirstResponder = nil
        UIApplication.shared.sendAction(#selector(findFristResponder(sender:)), to: nil, from: nil, for: nil)
        return UIResponder._currentFirstResponder
    }
    func findFristResponder(sender: AnyObject) {
        UIResponder._currentFirstResponder = self
    }
}
